In [1]:
print("Hello World")

Hello World


# Markdown Text in Jupyter Notebooks
A Markdown cell in a Jupyter Notebook cell can contain any text that's formatted with Markdown. Markdown is a relative simple language for marking up test with headings, lists, pictures videos, and more. Below is a video from YouTube that plays in this Notebook.
[![Free Video](https://img.youtube.com/vi/ZcH0fUn2AtM/0.jpg)]
(https://youtu.be/ZcH0fUn2AtM "Free Video")

In [1]:
x = -4
y = abs(x)
print(x)
print(y)

-4
4


In [3]:
x = 1.23456789098765432100000000000000000000000001
y = round(x,2)
print(x)
print(y)

1.2345678909876543
1.23


In [4]:
pi = 3.14159265358979
x = 128
y = -345.67890987
z = -999.9999
print(abs(z))
print(int(z))
print(int(abs(z)))
print(round(pi,4))
print(bin(x))
print(hex(x))
print(oct(x))
print(max(pi,x,y,z))
print(min(pi,x,y,z))
print(type(pi))
print(type(x))
print(type(str(y)))

999.9999
-999
999
3.1416
0b10000000
0x80
0o200
128
-999.9999
<class 'float'>
<class 'int'>
<class 'str'>


In [8]:
import math
z = 81
print(math.sqrt(81))

9.0


In [10]:
import math
pi = math.pi
e = math.e
tau = math.tau
x = 81
y = 7
z = -23234.5454
print(pi)
print(e)
print(tau)
print(math.sqrt(x))
print(math.factorial(y))
print(math.floor(z))
print(math.degrees(y))
print(math.radians(45))

3.141592653589793
2.718281828459045
6.283185307179586
9.0
5040
-23235
401.07045659157626
0.7853981633974483


# Formatting with f-Strings

f"Hello {username}"

In [14]:
username = "Alan"
print(f"Hello {username}")

Hello Alan


In [16]:
unit_price = 49.99
quantity = 30
print(f"Subtotal: ${quantity * unit_price}")

Subtotal: $1499.7


# Showing dollar amounts

Use commas to separate thousands from pennies

Subtotal: $1,499.70

:,

In [17]:
print(f"Subtotal: ${quantity * unit_price:,}")

Subtotal: $1,499.7


For 2 digit pennies use this .2f

In [19]:
print(f"Subtotal: ${quantity * unit_price:,.2f}")

Subtotal: $1,499.70


# Formatting percent numbers

sales_tax_rate = 0.065

In [22]:
sales_tax_rate = 0.065
print(f"Sales Tax Rate {sales_tax_rate}")

Sales Tax Rate 0.065


Use .2% toshow percentage in 2 decimal places

In [24]:
print(f"Sales Tax Rate {sales_tax_rate:.2%}")

Sales Tax Rate 6.50%


Using .1% to show number of digits

In [25]:
print(f"Sale Tax Rate {sales_tax_rate:.1%}")

Sale Tax Rate 6.5%


In [26]:
sales_tax_rate = 0.065
sample1 = f'Sales Tax Rate {sales_tax_rate:.2%}'
sample2 = f"Sales Tax Rate {sales_tax_rate:.2%}"
sample3 = f"""Sales Tax Rate {sales_tax_rate:.2%}"""
sample4 = f'''Sales Tax Rate {sales_tax_rate:.2%}''' 

print(sample1)
print(sample2)
print(sample3)
print(sample4)

Sales Tax Rate 6.50%
Sales Tax Rate 6.50%
Sales Tax Rate 6.50%
Sales Tax Rate 6.50%


# Making multiline format strings

In [27]:
user1 = "Alberto"
user2 = "Babs"
user3 = "Carlos"
output = f"{user1} \n{user2} \n{user3}"
print(output)

Alberto 
Babs 
Carlos


In [28]:
unit_price = 49.95
quantity = 32
sales_tax_rate = 0.065
subtotal = quantity * unit_price
sales_tax = sales_tax_rate * subtotal
total = subtotal + sales_tax
output=f"""
Subtotal:  ${subtotal:,.2f}
Sales Tax: ${sales_tax:,.2f}
Total:     ${total:,.2f}
"""
print(output)


Subtotal:  $1,598.40
Sales Tax: $103.90
Total:     $1,702.30



# Formatting width and alignment

Use f-strings with < for left-align, ^ for center-align and > for right-align

E.g :<20, :^20 and :>20

This means make this output 20 characters wide with content right, center or left-aligned 

In [32]:
unit_price = 49.95
quantity = 32
sales_tax_rate = 0.065
subtotal = quantity * unit_price
sales_tax = sales_tax_rate * subtotal
total = subtotal + sales_tax
output = f"""
Subtotal:  ${subtotal:>9,.2f}
Sales Tax: ${sales_tax:>9,.2f}
Total:     ${total:>9,.2f}
"""
print(output)


Subtotal:  $ 1,598.40
Sales Tax: $   103.90
Total:     $ 1,702.30



s_subtotal = "$" + f"{subtotal:,.2f}"

Subtotal: {s_subtotal:>9}

Use + to concatenate strings

In [34]:
# Numerical values
unit_price = 49.95
quantity = 32
sales_tax_rate = 0.065
subtotal = quantity * unit_price
sale_tax = sales_tax_rate * subtotal
total = subtotal + sales_tax

#Format amounts to show as string with leading dollar sign
s_subtotal = "$" + f"{subtotal:,.2f}"
s_sales_tax = "$" + f"{sales_tax:,.2f}"
s_total = "$" + f"{total:,.2f}"

# Output the string with dollar sign already attached
output = f"""
Subtotal:  {s_subtotal:>9}
Sales Tax: {s_sales_tax:>9}
Subtotal:  {s_total:>9}
"""
print(output)


Subtotal:  $1,598.40
Sales Tax:   $103.90
Subtotal:  $1,702.30



# Grappling with Weirder Numbers
## Binary, octal, and hexadecimal numbers

Dealing with base 2 = Binary, base 8 = Octal and base 16 = Hexadecimal numbers

In [35]:
print(hex(255))

0xff


In [36]:
x = 255
# Convert decimal to other number systems
print(bin(x))
print(oct(x))
print(hex(x))

# Show number in decimal number system (no conversion required)
print(0b11111111)
print(0o377)
print(0xff)

0b11111111
0o377
0xff
255
255
255


# Complex numbers

complex(real, imaginary)

In [37]:
z = complex(2, -3)
print(z)

(2-3j)


In [38]:
print(z.real)

2.0


In [39]:
print(z.imag)

-3.0


# Manipulating Strings

## Concatenating strings

Using + to concatinate strings

In [40]:
first_name = "Alan"
middle_name = "C"
last_name = "Simpson"
full_name = first_name+middle_name+last_name
print(full_name)

AlanCSimpson


Use "" to separate the strings

In [42]:
first_name = "Alan"
middle_name = "C"
last_name = "Simpson"
full_name = first_name + " " + middle_name + ". " + last_name
print(full_name)

Alan C. Simpson


# Getting the lenght of a string

Use a built-in tool or function len() to determine a lenght of a string.

In [45]:
s1 = ""
s2 = " "
s3 = "A B C"

print(len(s1))
print(len(s2))
print(len(s3))

0
1
5


# Working with common string operators

Playing around with string operators in Jupyter Notebook

In [46]:
s = "Abracadabra Hocus Pocus you're a turtle dove"
# Is there a lowercase letter t is contained in S?
print("t" in s)
# Is there an uppercase letter t is contained in S?
print("T" in s)
# Is there no uppercase T in s?S
print("T" not in s)
# Print 15 hyphens in a row
print("-" * 15)
# Print first character in string X
print(s[0])
# Print characters 33 - 39 from string x
print(s[33:39])
# Print every third character in s starting at zero
print(s[0:44:3])
# Print lowest character is s (a space is lower than the letter a)
print(min(s))
# Print the highest character is s
print(max(s))
# Where is the Uppercase P?
print(s.index("P"))
# Where is the first lowercase 0 in the latter half of string s
# Note that the returned value still starts counting form zero
print(s.index("o",22,44))
# How many lowercase letters a are in string s?
print(s.index("a"))

True
False
True
---------------
A
turtle
AadrHuPuy' tt v
 
y
18
25
3


Python offers two functions for working with ASCII. The ord() function takes a character
as input and returns the ASCII number of that character. The chr() does the opposite. You 
give it a number it returns the ASCII character for that number.

In [47]:
print(ord("A"))

65


In [48]:
print(chr(65))

A


# Manipulating strings with methods

string.methodname(params)

Playing around with Python 3 string functions

In [58]:
s1 = "There is no such word as scheedledorp"
s2="  a b c  "
s3="ABC"
s4="abc."
# Capitalize first letter, the rest lowercase
print(s3.capitalize())
# Count the number of spaces in s1
print(s1.count(" "))
#Find the dot in S4
print(s4.find("."))
#Is s2 all lowercase letters?
print(s2.islower())
#Convert s3 to all lowercase
print(s3.lower())
# String leading characters from s2
print(s2.lstrip())
# string leading and trailing characters from s2
print(s2.strip())
# Swap the case of letters in s1
print(s1.swapcase())
# Show s1 in title case (initial caps)
print(s1.title())
# Show s1 uppercase
print(s1.upper())


Abc
6
3
True
abc
a b c  
a b c
tHERE IS NO SUCH WORD AS SCHEEDLEDORP
There Is No Such Word As Scheedledorp
THERE IS NO SUCH WORD AS SCHEEDLEDORP


# Uncovering Dates and Times

Use the datetime module import datetime

Add alias (nickname) that's easier to type. E.g import datetime as dt

*datetime.date no time information
*datetime.time but no date
*datetime.datetime date and time

## Working with dates

#### Import the datetime module, nickname dt
import datetime as dt
#### Store today's date in a variable named today.
today = dt.date.today()
#### Store some other date in a variable called last_of_teens
last_of_teens = dt.date(2019,12,31)


In [62]:
# Import the datetime module give it an alias of dt
import datetime as dt
# Put today's date in today variable.
today = dt.date.today()
# Put another date in last_of_teens
last_of_teens = dt.date(2019,12,31)

# See what's in each variable.
print(today)
print(last_of_teens)

# Execute this code in the same cell
print(last_of_teens.month)
print(last_of_teens.day)
print(last_of_teens.year)

2023-02-04
2019-12-31
12
31
2019


### Using format string

In [67]:
print(f"{last_of_teens:%A, %B %d, %Y}")

Tuesday, December 31, 2019


### Show date in the mm/dd/yyyy format

In [70]:
todays_date = f"{today:%m/%d/%Y}"
print(todays_date)

02/04/2023


# Working with times

Basic syntax for time class

variable = datetime.time([hour, [minute,[microsecond]]])

In [73]:
midnight = dt.time()
print(midnight)

00:00:00


In [74]:
print(type(midnight))

<class 'datetime.time'>


In [75]:
almost_midnight = dt.time(23,59,59,999999)
print(almost_midnight)

23:59:59.999999


In [76]:
import datetime as dt
right_now = dt.datetime.now()
print(right_now)

2023-02-04 11:28:04.053234


datetime(year, month, day, hour, [minute, [second, [microsecond]]])

In [78]:
import datetime as dt
new_years_eve = dt.datetime(2019,12,31,23,59)
print(new_years_eve)

2019-12-31 23:59:00


### Calculating timespans

In [79]:
import datetime as dt
new_years_day = dt.date(2019,1,1)
memorial_day = dt.date(2019,5,27)
days_between = memorial_day - new_years_day

146 days, 0:00:00
<class 'datetime.timedelta'>

datetime.timedelta(days=, seconds=, microseconds=, milliseconds=, minutes;
hours=, week=)

In [81]:
import datetime as dt
new_year_day = dt.date(2019,1,1)
duration = dt.timedelta(days=146)
print(new_years_day + duration)

2019-05-27


In [82]:
import datetime as dt
memorial_day = dt.date(2019,5,27)
duration = dt.timedelta(days=146)
print(memorial_day - duration)

2019-01-01


In [83]:
import datetime as dt
start_time = dt.datetime(2019, 3, 31, 8, 0, 0)
finish_time = dt.datetime(2019, 3, 31, 14, 34, 45)
time_between = finish_time - start_time
print(time_between)
print(type(time_between))

6:34:45
<class 'datetime.timedelta'>


In [88]:
import datetime as dt
now = dt.datetime.now()
birthdatetime = dt.datetime(1995, 3, 31, 8, 26)
age = now - birthdatetime
print(age)
print(type(age))

10172 days, 3:54:23.794544
<class 'datetime.timedelta'>


In [89]:
import datetime as dt
today = dt.date.today()
birthdate = dt.date(2000, 12, 31)
delta_age = (today - birthdate)
print(delta_age)

8070 days, 0:00:00


In [90]:
delta_age = (today - birthdate)
days_old = delta_age.days
print(days_old, type(days_old))

8070 <class 'int'>


Use the floor division operator (//) rather than the  regular division (/)

In [91]:
years_old = days_old // 365
print(years_old)

22


In [93]:
import datetime as dt
# Today's date according to your computer
today = dt.date.today()

# Any birthdate expressed as year, month, day
birthdate = dt.date(1988, 3, 12)

# Duration between the dates as a timedelta
delta_age = (today - birthdate)

# Duration between the dates as a number (of days)
days_old = delta_age.days

# Floor divide days by 365 to  get the number of years
years = days_old // 365

# Days left over is remainder of days_lod divided by 365.
# Floor divide that remainder by 30 for approximate months
months = (days_old % 365) // 30

# Print in a format to your liking
print(f"You are {years} years and {months} months old")

You are 34 years and 11 months old


# Accounting for Time Zones

Using .now() and .utc_not()

In [95]:
# Get the datetime module and give it an alias
import datetime as dt

# Get the time from computer
here_now = dt.datetime.now()

# Get the time datetime right now
utc_now = dt.datetime.utcnow()

# Subtract to see difference 
time_difference = (utc_now - here_now)

# Show results
print(f"My time   : {here_now:%I:%M %p}")
print(f"UTC time  : {utc_now:%I:%M %p}")
print(f"Difference: {time_difference}")


My time   : 02:39 PM
UTC time  : 12:39 PM
Difference: -1 day, 22:00:00


Use dateutils and gettz (get timezome)

In [96]:
# import datetime and dateutil tz
import datetime as dt
from dateutil.tz import gettz

In [97]:
event = dt.datetime(2020,7,4,19,0,0)

#### Date and time for a scheduled event in multiple time zones.

In [98]:
# import datetime, give it an alias
import datetime as dt
# import timezone helpers from dateutil
from dateutil.tz import gettz

# UTC time right now.
utc=dt.datetime.now(gettz('Etc/UTC'))
print(f"{utc:%A %D %I:%M %p %Z}")

# USA Eastern time.
est = dt.datetime.now(gettz('America/New_York'))
print(f"{est:%A %D %I:%M %p %Z}")

# USA Central time.
cst=dt.datetime.now(gettz('America/Chicago'))
print(f"{cst:%A %D %I:%M %p %Z}")

# USA Mountain time.
mst=dt.datetime.now(gettz('America/Boise'))
print(f"{mst:%A %D %I:%M %p %Z}")

pst=dt.datetime.now(gettz('America/Los_Angeles'))
print(f"{pst:%A %D %I:%M %p %Z}")

Monday 02/06/23 01:01 PM UTC
Monday 02/06/23 08:01 AM EST
Monday 02/06/23 07:01 AM CST
Monday 02/06/23 06:01 AM MST
Monday 02/06/23 05:01 AM PST


#### The current date and time for five different time zones

syntex: name = event.astimezone(gettz("tzname))

syntex for built-in dateutil function .astimezome(gettz("tzname))

NB!!! Other authors recommend using arrow instead of dateutil when calculating timezomes.

In [100]:
# import datetime and dateutil tz
import datetime as dt
from dateutil.tz import gettz

# July 4 Event, 7:00 local time (no specific time zome).
event = dt.datetime(2020,7,4,19,0,0)
# Show local date and time
print("Local: " + f"{event:%D %I:%M %p %Z}" + "\n")

event_eastern = event.astimezone(gettz('America/New_York'))
print(f"{event_eastern:%D %I:%M %p %Z}")

event_central = event.astimezone(gettz('America/Chicago'))
print(f"{event_central:%D %I:%M %p %Z}")

event_mountain = event.astimezone(gettz('America/Denver'))
print(f"{event_mountain:%A %D %I:%M %p %Z}")

event_pacific = event.astimezone(gettz('America/Los_Angeles'))
print(f"{event_pacific:%D %I:%M %p %Z}")

event_utc = event.astimezone(gettz('Etc/UTC'))
print(f"{event_utc:%D %I:%M %p %Z}")



Local: 07/04/20 07:00 PM 

07/04/20 01:00 PM EDT
07/04/20 12:00 PM CDT
Saturday 07/04/20 11:00 AM MDT
07/04/20 10:00 AM PDT
07/04/20 05:00 PM UTC


# Chapter 2
### Controlling the Action

# Main operators for Controlling the Action

### Making Decisions with if

Syntax for if statement is: if condition: do this
do this no matter what

The first (do this) is executed only if the condition is true

In [107]:
# Indentation is very important here
sun = "down"
if sun == "down": 
    print("Good night!")
print("I am here")

Good night!
I am here


In [108]:
# Indentation is very important here
sun = "up"
if sun == "down":
    print("Good night!")
print("I an here")

I an here


if condition:
    do this
do this no matter what

In [112]:
sun = "down"
if sun == "down":
    print("Good night")
print("I am here")

Good night
I am here


In [118]:
total = 100
sales_tax_rate = 0.065
taxable = True     # Start with the capital Letter T instead of t
if taxable:        # Some people may say: if taxable == True:
    print(f"Subtotal : ${total:.2f}")
    sales_tax = total * sales_tax_rate
    print(f"Sales Tax: ${sales_tax:.2f}")
    total = total + sales_tax
print(f"Total   : ${total:.2f}")

Subtotal : $100.00
Sales Tax: $6.50
Total   : $106.50


#### When taxable is false, these are the results

In [119]:
total = 100
sales_tax_rate = 0.065
taxable = False     # Start with the capital Letter T instead of t
if taxable:        # Some people may say: if taxable == True:
    print(f"Subtotal : ${total:.2f}")
    sales_tax = total * sales_tax_rate
    print(f"Sales Tax: ${sales_tax:.2f}")
    total = total + sales_tax
print(f"Total   : ${total:.2f}")

Total   : $100.00


# Adding else to your if login

#### Print an initial greeting based on time of day

In [120]:
import datetime as dt
# Get the current date and time
now = dt.datetime.now()
# Make a decision based on hour
if now.hour < 12:
    print("Good morning")
else:
    print("Good afternoon")
print("I hope you are doing well!")

Good afternoon
I hope you are doing well!


# Handling multiple else's with elif

Syntax is:

In [121]:
light_color = "green"
if light_color == "green":
    print("Go")
elif light_color == "red":
    print("Stop")
print("This code executes no matter what")

Go
This code executes no matter what


In [122]:
light_color = "red"
if light_color == "green":
    print("Go")
elif light_color == "red":
    print("Stop")
print("This code executes no matter what")

Stop
This code executes no matter what


In [123]:
light_color = "yellow"
if light_color == "green":
    print("Go")
elif light_color == "red":
    print("Stop")
print("This code executes no matter what")

This code executes no matter what


In [124]:
light_color = "yellow"
if light_color == "green":
    print("Go")
elif light_color == "red":
    print("Stop")
else:
    print("Proceed with caution")
print("This code executes no matter what")

Proceed with caution
This code executes no matter what


# Ternary operations

##### Consider the following Javascript code

##### In Python you may re-write the above code as follows

In [127]:
age = 31
if age < 21:
    beverage = "milk"
elif age >= 21 and age < 80:
    beverage = "beer"
else:
    beverage = "prune juice"
print("Have a " + beverage)

Have a beer


##### Adding comments to the above code to make more readable

In [128]:
age = 31

if age < 21:
    # If under 21, no alcohol
    beverage = "milk"
    
elif age >= 21 and age < 80:
    # Ages 21 - 79, suggest beer
    beverage = "beer"
    
else:
    # If 80 or older, prune juice might be a good choice.
    beverage = "prune juice"
    
print("Have a " + beverage)

Have a beer


# Repeating a Process with for

### Decision making: 
for loops: it allows you to repeat a line of code, or several lines of code, 
    as many times as you like.

### Looping through numbers in a range

for loop sytax: 

In [131]:
for x in range(7): # This means range from 0 - 7
    print(x)
print("All done")

0
1
2
3
4
5
6
All done


In [133]:
# Here is the for loop with two numbers in range
for x in range(1, 10):
    print(x)
print("All done")

1
2
3
4
5
6
7
8
9
All done


# Looping through a string

Sytax for looping through a string:

In [134]:
for x in "snorkel":
    print(x)
print("Done")

s
n
o
r
k
e
l
Done


##### Using a variable name rather than a string in a for loop:

In [135]:
my_word = "snorkel"
for x in my_word:
    print(x)
print("Done")

s
n
o
r
k
e
l
Done


# Looping through a list

The syntax is as follows:

In [138]:
for x in ["The", "rain", "in", "Spain"]:
    print(x)
print("Done")

The
rain
in
Spain
Done


##### Assigning a list to a variable:

In [139]:
seven_dwarves = ["Happy", "Grumpy", "Sleepy", "Bashful", "Sneezy", "Doc", "Dopey"]
for dwarf in seven_dwarves:
    print(dwarf)
print("And Snow White too")

Happy
Grumpy
Sleepy
Bashful
Sneezy
Doc
Dopey
And Snow White too


# Bailing out of a loop

Use the break statement to force the loop stop

Syntax is:

In [140]:
answers = ["A", "C", "B", "D"]
for answer in answers:
    if answer == "":
        print("Incomplete")
        break
    print(answer)
print("Loop is done")

A
C
B
D
Loop is done


This loop jumps to an indented code as soon as it hits an empty string

In [142]:
answers = ["A", "C", "", "D"]
for answer in answers:
    if answer == "":
        print("Incomplete")
        break
    print(answer)
print("Loop is done")

A
C
Incomplete
Loop is done


# Looping with continue

Continue is the opposite of break. Continue makes the code execution jump to the top of the loop
and continue with the next item.

In [143]:
answers = ["A", "C", "", "D"]
for answer in answers:
    if answer == "":
        print("Incomplete")
        continue
    print(answer)
print("Loop is done")

A
C
Incomplete
D
Loop is done


# Nesting loops

In [150]:
# Outer loop
for outer in ["First", "Second", "Third"]:
    print(outer)
    # Inner loop
    for inner in range(3):
        print(inner + 1)
print("Both loops are done")
# Out of the loops here

First
1
2
3
Second
1
2
3
Third
1
2
3
Both loops are done


# Looping with while

Keep looping as long as some condition is true

Syntax for a while loop is as follows:

##### Looping while counter is less than 91

In [151]:
counter = 65
while counter < 91:
    print(str(counter) + "=" + chr(counter))
    counter += 1
print("all done")

65=A
66=B
67=C
68=D
69=E
70=F
71=G
72=H
73=I
74=J
75=K
76=L
77=M
78=N
79=O
80=P
81=Q
82=R
83=S
84=T
85=U
86=V
87=W
88=X
89=Y
90=Z
all done


In [1]:
# Removing the counter += will cause the following error
counter = 65
while counter < 91:
    print(str(counter) + "=" + chr(counter))
    counter += 1
print("all done")

65=A
66=B
67=C
68=D
69=E
70=F
71=G
72=H
73=I
74=J
75=K
76=L
77=M
78=N
79=O
80=P
81=Q
82=R
83=S
84=T
85=U
86=V
87=W
88=X
89=Y
90=Z
all done


##### A while loop with continue

In [4]:
import random 
print("Odd numbers")
counter = 0
while counter < 10:
    # Get a random number
    number = random.randint(1,999)
    if int(number / 2) == number / 2:
        # If it's an even number, don't print.
        continue
    # Otherwise, if it's odd, print it and increment the counter.
    print(number)
    # Increment the loop counter.
    counter += 1
print("Loop is done")

Odd numbers
209
391
413
125
535
831
789
609
237
887
Loop is done


# Breaking while loops with break

In [8]:
import random
print("Number that aren't evently divisible by 5")
counter = 0
while counter < 10:
    # Get a random number
    number = random.randint(1,999) 
    if int(number / 5) == number / 5:
        # If it's evently divisible by 5, bail out.
        break
    # Otherwise, print it and going for while.
    print(number)
    # Increment the loop counter.
    counter += 1
print("Loop is done")

Number that aren't evently divisible by 5
108
984
133
334
517
974
909
Loop is done


# Chapter 3
# Speeding Along with Lists and Tuples

#### Defining and  using Lists

Assign a name to a list using this: an = sign

In [9]:
# E.g: A list of integers
scores = [88, 92, 78, 90, 98, 84]

In [11]:
# E.g: A list of strings
students = ["Mark", "Amber", "Todd", "Anita", "Sandy"]

In [12]:
# Printing a students list
students = ["Mark", "Amber", "Todd", "Anita", "Sandy"]
print(students) # The following output is not an ideal way display the list.

['Mark', 'Amber', 'Todd', 'Anita', 'Sandy']


# Referencing list items by position

Refer to a list by its number:

Syntax:

listname = name of the list and x = the position of the item.

In [13]:
students = ["Mark", "Amber", "Todd", "Anita", "Sandy"]
print(students[0]) # here we display the name of the student at 0 index

Mark


In [16]:
scores = [88, 92, 78, 90, 84]
print(scores[4]) # here we display the score number of the score at 4th index

84


# Looping through a list

Use a for loop to loop through a list:

Syntax is as follows:

Use plural for list name: E.g scores and singular name for variable name: E.g score.

In [23]:
scores = [88, 92, 78, 90, 84]
for score in scores:
    print(score)
print("Done!")

88
92
78
90
84
Done!


# Seeing wether a list contains an item

In [25]:
students = ["Mark", "Amber", "Todd", "Anita", "Sandy"]

# Is Anita in the list?
has_anita = "Anita" in students
print(has_anita)

# Is Bob in the list?
has_bob = "Bob" in students
print(has_bob)

True
False


# Getting the lenght of a list

Use len() function to determine the lenght of a list

In [26]:
students = ["Mark", "Amber", "Todd", "Anita", "Sandy"]
print(len(students))

5


# Adding items to the end of a list

Use the .append() function to add items to the end of a list

In [27]:
# Create a list of strings (names)
students = ["Mark", "Amber", "Todd", "Anita", "Sandy"]

# Add the name Goober to the list
students.append("Goober") # This is one way of doing it!

new_student = "Amanda"
# Add what ever name is in new_student to the list.
students.append(new_student) # This is another way of doing it!

# Print the entire list.
print(students)

['Mark', 'Amber', 'Todd', 'Anita', 'Sandy', 'Goober', 'Amanda']


The following code won't add the Amber on the because the is already in the list

In [32]:
student_name = "Aneal"

# Add student_name but only if not already in the list.
if student_name in students:
    print(student_name + " already in the list")
else:
    students.append(student_name)
    print(student_name + " added to the list")

Aneal already in the list


In [33]:
print(students)

['Mark', 'Amber', 'Todd', 'Anita', 'Sandy', 'Goober', 'Amanda', 'Anele', 'Aneal']


# Inserting an item into a list

insert() allows you to add item in any position in a list

The syntax is:

In [36]:
# Create a list of strings (names)
students = ["Mark", "Amber", "Todd", "Anita", "Sandy"]

student_name = "Lupe"
# Add stuudent name to front of the list
students.insert(0, student_name)

# Show me the new list.
print(students)

['Lupe', 'Mark', 'Amber', 'Todd', 'Anita', 'Sandy']


# Changing an item in a list

The syntax is: 

In [2]:
# Create a list of strings (names).
students = ["Mark", "Amber", "Todd", "Anita", "Sandy"]
students[3] = "Hobart" # Here we replace Anita with Hobart.
print(students)

['Mark', 'Amber', 'Todd', 'Hobart', 'Sandy']


# Combining lists

Use the extend function to combine two lists.

In [3]:
# Create two lists of names.
list1 = ["Zara", "Lupe", "Hong", "Alberto", "Jake"]
list2 = ["Huey", "Dewey", "Louie", "Nader", "Bubba"]

# Add list2 names to list1
list1.extend(list2)

# Print list1.
print(list1)

['Zara', 'Lupe', 'Hong', 'Alberto', 'Jake', 'Huey', 'Dewey', 'Louie', 'Nader', 'Bubba']


# Removing list items

In [4]:
# Create a list of strings.
letters = ["A", "B", "C", "D", "E", "C"]

# Remove "C" from the list
letters.remove("C")

# Show me the new list.
print(letters)

['A', 'B', 'D', 'E', 'C']


In [5]:
# Create a list of strings.
letters = ["A", "B", "C", "D", "C", "E", "C"]

In [6]:
# Create a list of strings.
letters = ["A", "B", "C", "D", "E", "F", "G"]

# Remove the first item.
letters.pop(0)
# Remove the last item.
letters.pop()

# Show me the new list.
print(letters)

['B', 'C', 'D', 'E', 'F']


In [9]:
# Create a list of strings.
letters = ["A", "B", "C", "D", "E", "F", "G"]

# Make a copy of first list item then remove it from the list.
first_remove = letters.pop(0)
# Make a copy of last list item then remove it from the list.
last_remove = letters.pop()

# Show the new list.
print(letters)
# Show what's been removed.
print(first_remove + " and " + last_remove + " were removed from the list.")

['B', 'C', 'D', 'E', 'F']
A and G were removed from the list.


In [11]:
# Create a list of strings.
letters = ["A", "B", "C", "D", "E", "F", "G"]

# Remove item sub 2.
del letters[2]

print(letters)

['A', 'B', 'D', 'E', 'F', 'G']


Use del to delete the entire list.

The following code with produce an error

# Creating out a list

Use the clear() to delete items in the list but not the list. The result will produce []

In [13]:
# Create a list of strings.
letters = ["A", "B", "C", "D", "E", "F", "G"]

# Clear the list of all entries.
letters.clear()

# Show me the new list.
print(letters)

[]


# Counting how many times an item appears in a list

Use count() and str() to convert numbers to strings:

Syntax is:

In [15]:
# Create a list of strings.
grades = ["C", "B", "A", "D", "C", "B", "C"]

# Count the B's
b_grades = grades.count("B")

# Use a variable for value to count.
look_for = "C"
c_grades = grades.count(look_for)

print("There are " + str(b_grades) + " B grades in the list.")
print("There are " + str(c_grades) + " " + look_for + " grades in the list.")

# Count Fs too.
print("And there are " + str(grades.count("F")) + " F grades in the list.")

There are 2 B grades in the list.
There are 3 C grades in the list.
And there are 0 F grades in the list.


# Finding a list item's index

Python offers the .index() that returns a number indicating the position.

###### The above error can be fixed using an if statement

In [17]:
# Create a list of strings.
grades = ["C", "B", "A", "D", "C", "B", "C"]

# Decide what to look for 
look_for = "F"
# See if the item is in the list.
if look_for in grades:
    # If it's in the list, get and show the index.
    print(str(look_for) + " is at index " + str(grades.index(look_for)))
else:
    # If not in the list, don't even try for index number.
    print(str(look_for) + " isn't in the list.")

F isn't in the list.


# Alphabetizing and sorting lists

Syntax is:

##### Sorting strings and numbers

In [18]:
# Create a list of strings.
names = ["Zara", "Lupe", "Hong", "Alberto", "Jake", "Tyler"]
# Create a list of numbers.
numbers = [14, 0, 56, -4, 99, 56, 11.23]

# Sort the names list
names.sort()
# Sort the numbes list
numbers.sort()

# Show the results
print(names)
print(numbers)

['Alberto', 'Hong', 'Jake', 'Lupe', 'Tyler', 'Zara']
[-4, 0, 11.23, 14, 56, 56, 99]


##### Use a date module to sort dates

##### Sorting and displaying dates in a nice format

In [22]:
# Need this module for the dates.
import datetime as dt

# Create a list of dates, empty for starters
datelist = []
# Append dates one at time so code is easier to read.
datelist.append(dt.date(2020,12,31))
datelist.append(dt.date(2019,1,31))
datelist.append(dt.date(2018,2,28))
datelist.append(dt.date(2020,1,1))

# Sort the dates (earliest to latest) and show formatted.
datelist.sort()
for date in datelist:
    print(f"{date:%m/%d/%Y}")

02/28/2018
01/31/2019
01/01/2020
12/31/2020


Use reverse=True inside sort() parentheses and don't forget making the first an uppercase. 

In [23]:
# Need this modules for the dates.
import datetime as dt

# Create a list of strings.
names = ["Zara", "Lupe", "Hong", "Alberto", "Jake", "Tyler"]

# Create a list of numbers.
numbers = [14, 0, 56, -4, 99, 56, 11.23]

# Create a list of dates, empty for startes because code is long.
datelist = []
datelist.append(dt.date(2020,12,31))
datelist.append(dt.date(2019,1,31))
datelist.append(dt.date(2018,2,28))
datelist.append(dt.date(2020,1,1))

# Sort strings in reverse order (Z to A) and show.
names.sort(reverse=True)
print(names)
print() # This just adds a blank line to the output.

# Sort numbers in reverse order (largest to smallest) and show.
numbers.sort(reverse=True)
print(numbers)
print() # This just adds a blank line to the output.

# Sort the dates in reverse order (latest to earliest) and show formatted.
# Use a for loop.
datelist.sort(reverse=True)
for date in datelist:
    print(f"{date:%m/%d/%Y}")


['Zara', 'Tyler', 'Lupe', 'Jake', 'Hong', 'Alberto']

[99, 56, 56, 14, 11.23, 0, -4]

12/31/2020
01/01/2020
01/31/2019
02/28/2018


# Reversing a list

Use the .reverse method to reverse items in the list no matter their order.

In [24]:
# Create a list of strings.
names = ["Zara", "Lupe", "Hong", "Alberto", "Jake"]
# Reverse the list
names.reverse()
# Print the list
print(names)

['Jake', 'Alberto', 'Hong', 'Lupe', 'Zara']


# Copying a list

Use .copy() to make a copy and manipulate the copy not the original list.

In [25]:
# Create a list of strings.
names = ["Zara", "Lupe", "Hong", "Alberto", "Jake"]

# Make a copy of the list.
backward_names = names.copy()
# Reverse the copy
backward_names.reverse()

# Print the list
print(names)
print(backward_names)

['Zara', 'Lupe', 'Hong', 'Alberto', 'Jake']
['Jake', 'Alberto', 'Hong', 'Lupe', 'Zara']


# Methods for working with lists

# What's a Tuple and Who Cares?

tuple is an immutable list or an unchangable list.

You create a tuple in the same way you create a list except you don't use square brackets but you use parentheses ()

Syntax is:

In [26]:
prices = (29.95, 9.98, 4.95, 79.98, 2.95)

A tuple can't be modified but you can get the len of a tuple.

In [27]:
print(len(prices))

5


You can use .count() to count how many times an item appear in a tuple.

In [29]:
print(prices.count(4.95))

1


You can use in to see whether a value exist in a tuple.

In [30]:
print(4.95 in prices)

True


If an item exist in a tuple, you can get its index number.

In [32]:
look_for = 12345
if look_for in prices:
    position = prices.index(look_for)
else:
    position = -1
print(position)

-1


You can loop through a tuple and display items in a tuple using format string.

In [33]:
# Loop through and display each item in the tuple.
for price in prices:
    print(f"${price:.2f}")

$29.95
$9.98
$4.95
$79.98
$2.95


You cannot change the value of an item in tuple.
If you do try to change an item in a tuple you get the following error message: 
##### TypeError: 'tuple' object does not support item assignment

# Working with Sets

Use curly braces to define sets instead of square brackets for lists and parentheses for tuples. 

In [36]:
sample_set = {1.98, 98.9, 74.95, 2.5, 1, 16.3}

##### Sets are similar to lists and tuple in the following ways: The use len() to determine how many items in a set, they use in to determine whether item is in the set but you cannot get an item in a set based on its index number, nor can you change an item that is already in a set. You cannot change the order of items in a set either.

You can add a single item in a set using .add()

In [38]:
sample_set.add(11.23)
print(sample_set)

{16.3, 1.98, 98.9, 2.5, 1, 74.95, 11.23}


In [39]:
sample_set.add(13.23)
print(sample_set)

{16.3, 1.98, 98.9, 2.5, 1, 74.95, 11.23, 13.23}


##### You can add multiple items in a set using .update(). But the items you are adding should be defined as a list in square [] brackets.

In [41]:
sample_set.update([88, 123.45, 2.98])
print(sample_set)

{1.98, 2.5, 1, 2.98, 74.95, 11.23, 13.23, 16.3, 88, 98.9, 123.45}


You can copy a set but the items may not be the same as in the original set.

In [43]:
# Define a set named sample_set.
sample_set = {1.98, 98.9, 74.95, 2.5, 1, 16.3}
# Show the whole set
print(sample_set)
# Make a copy and shoe the copy.
ss2 = sample_set.copy()
print(ss2)

{16.3, 1.98, 98.9, 2.5, 1, 74.95}
{16.3, 1.98, 98.9, 2.5, 1, 74.95}


You can loop through a set and display its content formatted with with f-strings.
List and tuple are two of the most commonly used data structures. A fourth and a very widely used Python data structure is the data dictionary.

### Playing about with Python sets

In [44]:
# Define a set named sample_set
sample_set = {1.98, 98.9, 74.95, 2.5, 1, 16.3}
# Show the whole set
print(sample_set)

# Use len to get the lenght of a set.
print(len(sample_set))

# Use in to determine if the set contains a value
print(74.95 in sample_set)

# Use add() to add one item to a set.
sample_set.add(11.23)

# Use upadate() to add a [list] to a set
sample_set.update([88, 123.95, 2.98])

print("\nSample set after .add() and .update()")
print(sample_set)

# Loop through the set and print each item right-alighed and foramtted.
print("\nLoop through set and print each item formatted.")
for price in sample_set:
    print(f"{price:>6.2f}")

{16.3, 1.98, 98.9, 2.5, 1, 74.95}
6
True

Sample set after .add() and .update()
{1.98, 2.5, 1, 2.98, 74.95, 11.23, 16.3, 88, 98.9, 123.95}

Loop through set and print each item formatted.
  1.98
  2.50
  1.00
  2.98
 74.95
 11.23
 16.30
 88.00
 98.90
123.95


# Chapter 4  

## Cruising Massive Data with Dictionaries

A data dictionary with keys in the left column, values in the right.

A data dictionary with lists as values.

A data dictionary for one employee.

The key name is more descriptive than using an index thats based on position.

# Creating a Data Dictionary

The syntax is as follows:

Or you can use the below option for readability

In [3]:
people = {
    'htanaka': 'Haru Tanaka',
    'ppatel' : 'Priya Pate',
    'bagarcia': 'Benjamin Alberto Garcia',
    'zmin': 'Zhang Min',
    'afarooqi': 'Ayesha Farooqi',
    'hajackson': 'Hanna Jackson',
    'papatel':  'Praytyush Aarav Patel',
    'hrjackson': 'Henry Jackson'
}

# Accessing dictionary data

In [4]:
print(people)
{
    'htanaka': 'Haru Tanaka',
    'ppatel' : 'Priya Pate',
    'bagarcia': 'Benjamin Alberto Garcia',
    'zmin': 'Zhang Min',
    'afarooqi': 'Ayesha Farooqi',
    'hajackson': 'Hanna Jackson',
    'papatel':  'Praytyush Aarav Patel',
    'hrjackson': 'Henry Jackson'
}

{'htanaka': 'Haru Tanaka', 'ppatel': 'Priya Pate', 'bagarcia': 'Benjamin Alberto Garcia', 'zmin': 'Zhang Min', 'afarooqi': 'Ayesha Farooqi', 'hajackson': 'Hanna Jackson', 'papatel': 'Praytyush Aarav Patel', 'hrjackson': 'Henry Jackson'}


{'htanaka': 'Haru Tanaka',
 'ppatel': 'Priya Pate',
 'bagarcia': 'Benjamin Alberto Garcia',
 'zmin': 'Zhang Min',
 'afarooqi': 'Ayesha Farooqi',
 'hajackson': 'Hanna Jackson',
 'papatel': 'Praytyush Aarav Patel',
 'hrjackson': 'Henry Jackson'}

Use this syntax if you want a specific item in the dictionary.

E.g.

#### Printing the value of the zmin key in the people dictionary

In [2]:
# Make a data dictionary named people
people = {
    'htanaka': 'Haru Tanaka',
     'ppatel': 'Priya Pate',
     'bagarcia': 'Benjamin Alberto Garcia',
     'zmin': 'Zhang Min',
     'afarooqi': 'Ayesha Farooqi',
     'hajackson': 'Hanna Jackson',
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}
print(people['zmin'])

Zhang Min


In [3]:
# Make a data dictionary named people
people = {
    'htanaka': 'Haru Tanaka',
     'ppatel': 'Priya Pate',
     'bagarcia': 'Benjamin Alberto Garcia',
     'zmin': 'Zhang Min',
     'afarooqi': 'Ayesha Farooqi',
     'hajackson': 'Hanna Jackson',
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}

# Look for a person
person = 'zmin'
print(people[person])

Zhang Min


In [8]:
# Make a data dictionary named people
people = {
    'htanaka': 'Haru Tanaka',
     'ppatel': 'Priya Pate',
     'bagarcia': 'Benjamin Alberto Garcia',
     'zmin': 'Zhang Min',
     'afarooqi': 'Ayesha Farooqi',
     'hajackson': 'Hanna Jackson',
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}
# Look for a person
person = 'hrjackson'
print(people[person])

Henry Jackson


##### Python's way of saying there's no scheedledorp 

In [9]:
# Make a data dictionary named people
people = {
    'htanaka': 'Haru Tanaka',
     'ppatel': 'Priya Pate',
     'bagarcia': 'Benjamin Alberto Garcia',
     'zmin': 'Zhang Min',
     'afarooqi': 'Ayesha Farooqi',
     'hajackson': 'Hanna Jackson',
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}
# Look for a person
person = 'scheedledorp'
print(people[person])

KeyError: 'scheedledorp'

### Getiing the lenght of a dictionary

##### Syntax: 

In [1]:
people = {
    'htanaka': 'Haru Tanaka',
     'ppatel': 'Priya Pate',
     'bagarcia': 'Benjamin Alberto Garcia',
     'zmin': 'Zhang Min',
     'afarooqi': 'Ayesha Farooqi',
     'hajackson': 'Hanna Jackson',
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}
# Count the number of key: value pairs and put in a variable.
howmany = len(people)
# Show how many.
print(howmany)

8


## Seeing whether a key exists in a dictionary

In [3]:
# Make a data dictionary named people
people = {
    'htanaka': 'Haru Tanaka',
     'ppatel': 'Priya Pate',
     'bagarcia': 'Benjamin Alberto Garcia',
     'zmin': 'Zhang Min',
     'afarooqi': 'Ayesha Farooqi',
     'hajackson': 'Hanna Jackson',
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}

# Is there an hajackson in the people dictionary?
print('hajackson' in people)

# Is there an scheedledorp in the people dictionary?
print('scheedledorp' in people)


True
False


## Getting dictionary data with get()

##### Syntax

##### Python's way of saying there is no scheedlodorp in the dictionary

In [4]:
# Make a data dictionary named people
people = {
    'htanaka': 'Haru Tanaka',
     'ppatel': 'Priya Pate',
     'bagarcia': 'Benjamin Alberto Garcia',
     'zmin': 'Zhang Min',
     'afarooqi': 'Ayesha Farooqi',
     'hajackson': 'Hanna Jackson',
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}

# look for a person.
person = 'scheedledorp'
print(people.get(person))

None


##### Lets make present something better than none

In [6]:
# Make a data dictionary named people
people = {
    'htanaka': 'Haru Tanaka',
     'ppatel': 'Priya Pate',
     'bagarcia': 'Benjamin Alberto Garcia',
     'zmin': 'Zhang Min',
     'afarooqi': 'Ayesha Farooqi',
     'hajackson': 'Hanna Jackson',
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}

# look for a person.
person = 'scheedledorp'
print(people.get('scheedledorp','Unbeknownst to this dictionary'))

Unbeknownst to this dictionary


## Changing the value of a key

##### Syntax:

In [7]:
# Print hajackson's current value.
print(people['hajackson'])

# Change the value of hajackson's key.
people['hajackson'] = "Hanna Jackson-Smith"

# Print the hajackson key to verify that the value has changed.
print(people['hajackson'])

Hanna Jackson
Hanna Jackson-Smith


## Adding or changing dictionary data

##### Use the update method

In [11]:
# Make a data dictionary named people
people = {
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}

# Change the value of the hr jackson key.
people.update({'hr jackson':'Henrietta Jackson'})
print(people)

# Update the dictionary with a new property: value pair.
people.update({'wwigging':'Wanda Wiggings'})
print(people)

{'papatel': 'Praytyush Aarav Patel', 'hrjackson': 'Henry Jackson', 'hr jackson': 'Henrietta Jackson'}
{'papatel': 'Praytyush Aarav Patel', 'hrjackson': 'Henry Jackson', 'hr jackson': 'Henrietta Jackson', 'wwigging': 'Wanda Wiggings'}


In [14]:
# Make a data dictionary named people
people = {
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}

# Change the value of the hr jackson key.
#people.update({'hr jackson':'Henrietta Jackson'})
#print(people)

# Update the dictionary with a new property: value pair.
people.update({'wwigging':'Wanda Wiggings'})
print(people)

# Show what's in the data dictionary now.
for person in people.keys():
    print(person + " = " + people[person])

{'papatel': 'Praytyush Aarav Patel', 'hrjackson': 'Henry Jackson', 'wwigging': 'Wanda Wiggings'}
papatel = Praytyush Aarav Patel
hrjackson = Henry Jackson
wwigging = Wanda Wiggings


## Looping through a Dictionary

In [15]:
for person in people:
    print(person)

papatel
hrjackson
wwigging


In [16]:
for person in people:
    print(people[person])

Praytyush Aarav Patel
Henry Jackson
Wanda Wiggings


In [17]:
for person in people.values():
    print(person)

Praytyush Aarav Patel
Henry Jackson
Wanda Wiggings


##### Looping through a dictionary with items

In [18]:
# Make a data dictionary named people
people = {
    'htanaka': 'Haru Tanaka',
     'ppatel': 'Priya Pate',
     'bagarcia': 'Benjamin Alberto Garcia',
     'zmin': 'Zhang Min',
     'afarooqi': 'Ayesha Farooqi',
     'hajackson': 'Hanna Jackson',
     'papatel': 'Praytyush Aarav Patel',
     'hrjackson': 'Henry Jackson'
}

# Looping through .items to gte the key and the value.
for key, value in people.items():
    # Show the and value with = in betwwen.
    print(key, " = ", value)

htanaka  =  Haru Tanaka
ppatel  =  Priya Pate
bagarcia  =  Benjamin Alberto Garcia
zmin  =  Zhang Min
afarooqi  =  Ayesha Farooqi
hajackson  =  Hanna Jackson
papatel  =  Praytyush Aarav Patel
hrjackson  =  Henry Jackson


## Data Dictionary Methods

##### Copying a Dictionary

Use this syntax: newdictionaryname = dictionary

In [19]:
# Define a dictionary
people = {
    'htanaka': 'Haru Tanaka',
    'zmin': 'Zhang Min',
    'afarooqi': 'Ayesha Farooqi',
}

# Make a copy of the people dictionay and put it in peep 2.
peep2 = people.copy()

# Show what is in both dictionaries
print(people)
print(peep2)




{'htanaka': 'Haru Tanaka', 'zmin': 'Zhang Min', 'afarooqi': 'Ayesha Farooqi'}
{'htanaka': 'Haru Tanaka', 'zmin': 'Zhang Min', 'afarooqi': 'Ayesha Farooqi'}


## Deleting Dictionary Items

In [20]:
# Define a dictionary named people.
people = {
    'htanaka': 'Haru Tanaka',
    'zmin': 'Zhang Min',
    'afarooqi': 'Ayesha Farooqi',
}

# Show original people dictionary.
print(people)

# Remove zmin from the dictionary.
del people["zmin"]

# Show what's in people now.
print(people)

{'htanaka': 'Haru Tanaka', 'zmin': 'Zhang Min', 'afarooqi': 'Ayesha Farooqi'}
{'htanaka': 'Haru Tanaka', 'afarooqi': 'Ayesha Farooqi'}


##### To remove all key value pairs from a dictionary, without deleting the entire dictionary. 

In [21]:
# Define a dictionary named people.
people = {
    'htanaka': 'Haru Tanaka',
    'zmin': 'Zhang Min',
    'afarooqi': 'Ayesha Farooqi',
}

# Show original people dictionary.
print(people)

# Remove all data from the dictionary.
people.clear()

# Show what's in people now.
print(people)

{'htanaka': 'Haru Tanaka', 'zmin': 'Zhang Min', 'afarooqi': 'Ayesha Farooqi'}
{}


## Using pop() with Data Dictionaries 

In [22]:
# Define a dictionary named people.
people = {
    'htanaka': 'Haru Tanaka',
    'zmin': 'Zhang Min',
    'afarooqi': 'Ayesha Farooqi',
}

# Show original people dictionary.
print(people)

# Pop zmin from the dictionary, store its value  in adios variable
adios = people.pop("zmin")

# Print the contents of adios and people.
print(adios)
print(people)

{'htanaka': 'Haru Tanaka', 'zmin': 'Zhang Min', 'afarooqi': 'Ayesha Farooqi'}
Zhang Min
{'htanaka': 'Haru Tanaka', 'afarooqi': 'Ayesha Farooqi'}


In [23]:
# Define a dictionary named people.
people = {
    'htanaka': 'Haru Tanaka',
    'zmin': 'Zhang Min',
    'afarooqi': 'Ayesha Farooqi',
}

# Show original people dictionary.
print(people)

# Pop zmin from the dictionary, store its value  in adios variable
adios = people.popitem()

# Print the contents of adios and people.
print(adios)
print(people)

{'htanaka': 'Haru Tanaka', 'zmin': 'Zhang Min', 'afarooqi': 'Ayesha Farooqi'}
('afarooqi', 'Ayesha Farooqi')
{'htanaka': 'Haru Tanaka', 'zmin': 'Zhang Min'}


## Fun with Multi-key Dictionaries

In [27]:
employee = {
    'name' : 'Haru Tanaka',
    'year_hired' : 2005,
    'dob' : '11/23/1987',
    'has_laptop' : False
}

In [28]:
product = {
    'name' : 'Ray-Ban Wayfarer Sunglasses',
    'unit_price' : 112.99,
    'taxable' : True,
    'in_stock' : 10
}

In [29]:
product = {
    'name' : 'Ray-Ban Wayfarer Sunglasses',
    'unit_price' : 112.99,
    'taxable' : True,
    'in_stock' : 10,
    'models' : ['Black', 'Tortose']
}

In [30]:
print(product['name'])
print(product['unit_price'])
print(product['taxable'])
print(product['in_stock'])
print(product['models'])

Ray-Ban Wayfarer Sunglasses
112.99
True
10
['Black', 'Tortose']


In [32]:
product = {
    'name' : 'Ray-Ban Wayfarer Sunglasses',
    'unit_price' : 112.99,
    'taxable' : True,
    'in_stock' : 10,
    'models' : ['Black', 'Tortose']
}

print('Name:      ', product['name'])
print('Price:     ', product['unit_price'])
print('Taxable:   ', product['taxable'])
print('In Stock:  ',product['in_stock'])
print('Models: ')
for model in product['models']:
    print(" " * 10 + model)

Name:       Ray-Ban Wayfarer Sunglasses
Price:      112.99
Taxable:    True
In Stock:   10
Models: 
          Black
          Tortose


## Using the mysterious fromkeys and set default methods

In [33]:
DWC001 = dict.fromkeys(['name', 'unit_price', 'taxable', 'in_stock', 'models'])
print(DWC001)

{'name': None, 'unit_price': None, 'taxable': None, 'in_stock': None, 'models': None}


In [34]:
# Create a generic dictionary for products name product.
product = {
    'name' : '',
    'unit_price': 0,
    'taxable': True,
    'in_stock': 0,
    'models': []
}

# Create a dictionary names DWC001 that has same keys as product.
DWC001 = dict.fromkeys(product.keys())

# Show what's in the new dictionary.
print(DWC001)

{'name': None, 'unit_price': None, 'taxable': None, 'in_stock': None, 'models': None}


# Experimenting with fromkeys and setdefault

In [36]:
# Create a generic dictonary for products name product.
product = {
    'name': '',
    'unit_price' : 0, 
    'taxable': True,
    'in_stock': 0,
    'models': []
    
}

# Create a dictionary for product SKU # DWC001
DWC001 = dict.fromkeys(product.keys())
DWC001.setdefault('taxable', True)
DWC001.setdefault('models', [])
DWC001.setdefault('reorder_point', 100)

# Show what's in the new dictionary.
print("Dictionary after fromkeys() and setdefault()")
print(DWC001)

# Change the taxable field form None to True.
print("\nDictionary after fromkeys() and setdefault()")
DWC001['taxable'] = True

# Print the dictionary after changing taxable to True
print(DWC001)

Dictionary after fromkeys() and setdefault()
{'name': None, 'unit_price': None, 'taxable': None, 'in_stock': None, 'models': None, 'reorder_point': 100}

Dictionary after fromkeys() and setdefault()
{'name': None, 'unit_price': None, 'taxable': True, 'in_stock': None, 'models': None, 'reorder_point': 100}


# Nesting Dictionaries

# Multiple product dictionaries contained within a larger products dictionary

In [37]:
# Create a generic dictioinary to contain multiple product dictionaries.
products = {
    'RB00111':{'name' : 'Rayban Sunglasses', 'price':112.98, 'models': ['black', 'tortoise']},
    'DWC0317':{'name' : 'Drone with Camera', 'price':72.95, 'models': ['white', 'black']},
    'MTS0540':{'name' : 'T-Shirt', 'price':2.95, 'models': ['small', 'medium', 'large']},
    'ECD2989':{'name' : 'Echo Dot', 'price':29.99, 'models':[]},
}

# Printing data dictionaries formatted into rows and columns.

In [38]:
# Create a generic dictioinary to contain multiple product dictionaries.
products = {
    'RB00111':{'name' : 'Rayban Sunglasses', 'price':112.98, 'models': ['black', 'tortoise']},
    'DWC0317':{'name' : 'Drone with Camera', 'price':72.95, 'models': ['white', 'black']},
    'MTS0540':{'name' : 'T-Shirt', 'price':2.95, 'models': ['small', 'medium', 'large']},
    'ECD2989':{'name' : 'Echo Dot', 'price':29.99, 'models':[]},
}
print(f"{'ID':<6} {'Name':<17} {'Price':>8} {'Models'}")
print('-' * 60)
# Loop through each dictionary in the products dictionary
for oneproduct in products.keys():
    # Get the id of one product.
    id = oneproduct
    # Get the name of one product.
    name = products[oneproduct]['name']
    # Get the unit price of one product and format with $
    unit_price = '$' + f"{products[oneproduct]['price']:,.2f}"
    # Create and empty string variable named models
    models = ''
    # Loop through the models list and tack onto models
    # one item from the list followed by by a comma and a space.
    for m in products[oneproduct]['models']:
        models += m + ', '
    # If the models variable is more than two characters in lenght,
    # Peel off the last two characters (last comma and space)
    if len(models) > 2:
        models = models[:-2]
    else:
        models = "<none>"
    # Print all the variables with a neat f-string.
    print(f"{id:<6} {name:<17} {unit_price:>8} {models}")

ID     Name                 Price Models
------------------------------------------------------------
RB00111 Rayban Sunglasses  $112.98 black, tortoise
DWC0317 Drone with Camera   $72.95 white, black
MTS0540 T-Shirt              $2.95 small, medium, large
ECD2989 Echo Dot            $29.99 <none>


# Chapter 5

# Wrangling Bigger Chunks of Code

#### Use a def key word to create a function

Syntax: 

In [3]:
def hello():
    print('Hello World')  # This line of code will run until the function is called.
hello()                   # Now we calling this function.   

Hello World


## Commenting a function

In [4]:
def hello():    # Practice function
    """"A docstring describing the function"""
    print('Hello')

## Passing information to a Function

In [7]:
def hello(x):     # Practice function
    """A docstring describing the function"""
    print('Hello ' + x)

In [9]:
def hello(user_name):    # practice function
    """A docstring describing the function"""
    print('Hello ' + user_name)

In [12]:
def hello(user_name):    # practice function
    """A docstring describing the function"""
    print('Hello ' + user_name)
hello()

TypeError: hello() missing 1 required positional argument: 'user_name'

In [13]:
def hello(user_name):     # Practice function
    """A docstring describing the function"""
    print('Hello ' + user_name)
hello('Alan')

Hello Alan


## Passing data to a functioin via a variable

In [14]:
def hello(user_name):    # Practice function
    """A docstring describing the function"""
    print('Hello ' + user_name)
    
# Put a string in a variable named this_person
this_person = 'Alan'
# Pass that variable name to the function
hello(this_person)


Hello Alan


# Defining optional parameters with defaults

Syntax:

### you could rewrite our sample hello() function with a default value, like this:

### Calling a function with and without parameter

In [20]:
def hello(user_name = 'nobody'):
    """A docstring describing the function"""
    print('Hello ' + user_name)
    
hello('Alan')
hello()

Hello Alan
Hello nobody


# Passing multiple values to a function

In [21]:
def hell(fname, lname, datestring):      # Practice fucntion
    """A docstring describing the function"""
    print('Hello ' + fname + '' + lname)
    print('The date is ' + datestring)

#### when calling the function, you need to pass in three values, like this:

In [23]:
def hello(fname, lname, datestring):      # Practice function
    """A docstring describing the function"""
    msg = "Hello " + fname + " " + lname
    msg += " you mentioned " + datestring
    print(msg)
    
hello('Alan', 'Simpson', '12/31/2019')

Hello Alan Simpson you mentioned 12/31/2019


### Make the optional parameters the last ones when using multiple parameters

In [28]:
def hello(fname, lname, datestring = ''):
    msg = 'Hello ' + fname + ' ' + lname
    if len(datestring) > 0:
        msg += ' you mentioned ' + datestring
    print(msg)
    
hello('Alan', 'Simpson', '12/31/2019')
hello('Sammy', 'Scmeedledorp')

Hello Alan Simpson you mentioned 12/31/2019
Hello Sammy Scmeedledorp


# Using keyword arguments (kwargs)

### You can tell the function what’s what by using the syntax parameter = value in the code that’s calling the function

##### It works if you pass in variables too. Again, the order doesn’t matter. Here’s another way to do this:

### Calling a function with keyword arguments (kwargs)

In [33]:
def hello(fname, lname, datestring):      # Practice function
    """A docstring describing the function"""
    msg = "Hello " + fname + " " + lname
    msg += "you mentioned " + datestring
    print(msg)
    
# Pass in in literal kwargs (identify each by parameter name)
hello(datestring='12/30/2019', lname='Simpson', fname='Alan')

# Pass in in kwargs from variables (identity each by parameter name)
appt_date = '12/30/2019'
last_name = 'Janda'
first_name = 'Kylie'
hello(datestring=appt_date, lname=last_name, fname=first_name)

Hello Alan Simpsonyou mentioned 12/30/2019
Hello Kylie Jandayou mentioned 12/30/2019


In [38]:
def alphabetize(original_list = []):
    """Pass any list in square brackets, displays a string with items sorted"""
    # Inside the function make a working copy of the list passed in.
    sorted_list = original_list.copy()
    # Sort the working copy.
    sorted_list.sort()
    # Make a new empty string for output
    final_list = ''
    # Loop through sorted list and append name and comma and space.
    for name in sorted_list:
        final_list += name + ', '
    # Knock of last comma space if final list is long enough
    final_list = final_list[:-2]
    # Print the alphabetized list.
    print(final_list)
    


##### Because the function can’t alter the list directly, it first makes a copy of the origi-nal list (the one that was passed in) in a new list called sorted_list, with this line of code:

##### At this point, the sorted_list isn’t really sorted, it’s still just a copy of the origi-nal. The next line of code does the actual sorting:

##### This loop loops through the sorted list and adds each item from the list, separated by a comma and a space, to that final_list string:

###### this statement removes those last two charac-ters, assuming the list is at least two characters in length:

##### To call the function, you can pass a list right inside the parentheses of the func-tion, like this:

##### As always, you can also pass in the name of a variable that already contains the list, as in this example.

#### Either way, the function shows those names in alphabetical order, as follows.

# Passing in an arbitrary number of arguments

In [39]:
def sorter(*args):
    """Pass in any number of arguments separated by commas inside the function, they treated as a tuple named args"""
    # The passed-in
    # create a list from the passed-in tuple
    newlist = list(args)
    # Sort and show the list.
    newlist.sort()
    print(newlist)
    
sorter(1, 0.001, 100000, -900, 2)   

[-900, 0.001, 1, 2, 100000]


# Returning Values from functions

In [41]:
def alphabetized(original_list=[]):
    """Pass any list in square brackets, display a string with items sorted"""
    
    # Inside a function make a working copy of the list passed in.
    sorted_list = original_list.copy()
    
    # Sort the working copy
    sorted_list.copy()
    
    # Make a new empty string for output
    final_list = ''
    
    # Loop through sorted list and append name and comma and space.
    for name in sorted_list:
        final_list += name + ', '
        
    # Knock of last comma space if final list is long enough
    final_list = final_list[:-2]
    
    # Return the alphabetized list.
    return final_list

## The final print statement displays whatever is in the alpha_list variable:

### Printing a string returned by the alphabetize() function.

In [42]:
def alphabetize(original_list=[]):
    """Pass any list in square brackets, displays with items sorted"""
    
    # Inside the function make a working copy of the list passed in.
    sorted_list = original_list.copy()
    
    # Sort the working copy.
    sorted_list.sort()
    
    # Make a new empty string for output
    final_list = ''
    
    # Loop through sorted list and append name and comma and space
    for name in sorted_list:
        final_list += name + ', '
        
    # Knock of last comma space if final list is long enough
    final_list = final_list[:-2]
    
    # Print the alphabetized list
    print(final_list)
    
names = ['McMullen', 'Keaser', 'Maier', 'Wilson', 'Yudt', 'Gallagher', 'Jacobs']
alphabetize(names)

Gallagher, Jacobs, Keaser, Maier, McMullen, Wilson, Yudt


# Unmasking Anonymous Functions

Syntax:

##### Names that starts with uppercase and those that starts with lowercase letters:

In [1]:
names = ['Adams', 'Ma', 'diMeola', 'Zandusky']
names.sort()
print(names)

['Adams', 'Ma', 'Zandusky', 'diMeola']


In [3]:
def lowercaseof(anystring):
    """Convert string to all lowercase"""
    return anystring.lower()

print(lowercaseof('Zandusky'))

zandusky


In [4]:
def lowercaseof(anystring):
    """ Converts string to all lowercase """
    return anystring.lower()

names = ['Adams','Ma','diMeola','Zandusky']
names.sort(key=lowercaseof)

In [ ]:
lambda parameters: expression

In [5]:
lambda anystring: anystring.lower()

<function __main__.<lambda>(anystring)>

### Using a lambda expression as a sort key

In [6]:
names = ['Adams','Ma','diMeola','Zandusky']
names.sort(key = lambda anystring: anystring.lower())
print(names)

['Adams', 'diMeola', 'Ma', 'Zandusky']


In [8]:
currency = lambda n: f"${n:,.2f}"

In [9]:
percent = lambda n: f"n:.2%"

In [10]:
# Show number in currency format.
def currency(n):
    return f"${n:,.2f}"


In [11]:
def percent(n):
# Show number in percent format.
    return f"{n:.2%}"

### Two functions for formatting numbers

In [13]:
# Show number in currency format.
currency = lambda n: f"${n:,.2f}"
# Show number in percent format.
percent = lambda n: f"{n:.2%}"

#Test currency function
print(currency(99))
print(currency(123456789.09876543))

# Test percent function
print(percent(0.065))
print(percent(.5))

$99.00
$123,456,789.10
6.50%
50.00%


### Two for formatting numbers with fixed width

In [15]:
# Show numbers in currency, specified width
def currency(n, w=15):
    """Show in currency format, width = 15 or width of your choosing"""
    s = f"${n:,.2f}"
    # Pad left of output with spaces to width of w.
    return s.rjust(w)

# Show numbers in percent format, specify width.
def percent(n, w=15):
    """Show in percent format, width = 15 or width of your choosing"""
    # Show number in percent format.
    s = f"{n:.1%}"
    # Pad left of output with spaces to width of w.
    return s.rjust(w)

In [16]:
print(currency(9999))

      $9,999.00


In [17]:
print(currency(9999.20))

      $9,999.20


# Chapter 6
# Doing python with Class

# Mastering Classes and Objects

#### Creating a Class

In [ ]:
# Define a new class name Member
class Member:
    """ Create a new member. """

In [19]:
# Define a new class name Member.
class Member:
    pass

# How a class create an Instance

### The syntax for creating an init method is:

# Giving an Object Its Attributes

# Creating an instance from a class

# Creating a member from the Member class in a Jupyter cell

In [26]:
# Define a class name Member
class Member:
    """ Create a new member. """
    def __init__(self, uname, fname):
        # Define attributes and give them values.
        self.username = uname
        self.fullname = fname
        
# The class end at the first un-indented line.

# Create an instance of a Member class named new_guy
new_guy = Member('Rambo', 'Rocco Moe')

# See what's in the instance, as well as its individual properties.
print(new_guy)
print(new_guy.username)
print(new_guy.fullname)
print(type(new_guy))

Rambo
Rocco Moe
<class '__main__.Member'>


# The Member class with username and fullname for parameters and attributes

In [27]:
# Define a new class name Member.
class Member:
    """ Create a new member. """
    def __init__(self, username, fullname):
        # Define attributes and give them values.
        self.username = username
        self.fullname = fullname
        
# The class ends at the first un-indented line.

# Create an instance of the Member class named new_guy
new_guy = Member('Rambo', 'Rocco Moe')

# See what's in the instance, as well as its individual properties.
print(new_guy)
print(new_guy.username)
print(new_guy.fullname)
print(type(new_guy))

Rambo
Rocco Moe
<class '__main__.Member'>


# Changing the value of an attribute

In [28]:
new_guy.username = "Princess"

# Defining attributes with default values

# Changing the value of an object's attribute

In [32]:
# Define a new class name Member.
class Member:
    """ Create a new member. """
    def __init__(self, username, fullname):
        # Define attributes and give them values.
        self.username = username
        self.fullname = fullname
    
# The class end at the first un-indented line.

# Create an instance of the Member class named new_guy
new_guy = Member('Rambo', 'Rocco Moe')

# See what's in the instance, as well as its individual properties.
print(new_guy.username)
print(new_guy.fullname)
print() # This just prints a blanks line.

# Change new_guy's user name then print both attributes again.
new_guy.username = "Princess"
print(new_guy.username)
print(new_guy.fullname)

Rambo
Rocco Moe

Princess
Rocco Moe


### If you're going to be doing anything with dates and times, you'll want to import the datetime module

In [35]:
import datetime as dt

# Define a new class name Member.
class Member:
    """ Create a new member. """
    def __init__(self, username, fullname):
        # Define attributes and give them values.
        self.username = username
        self.fullname = fullname
        
        # Default date_joined to toady's date.
        self.date_joint = dt.date.today()
        # Set is active to True initially
        self.is_active = True

#### Change any other attribute's value using the syntax:

# Giving a Class Methods

#### To define what the methods will be in your class, use this syntax for each method:

In [36]:
# Define methods as functions, use self for "this" object.
def show_date_joined(self):
    return f"{self.fullname} joined on {self.date_joined:%m/%d/%y}"

# Changing the value of an object's attributes

In [42]:
import datetime as dt

# Define a new class name Member
class Member:
    """ Create a new member. """
    def __init__(self, username, fullname):
        # Define attribute and give them values
        self.username = username
        self.fullname = fullname
        # Default date_joined to today's date
        self.date_joined = dt.date.today()
        # Set is active to True initially.
        self.is_active = True
        
    # Member for each instance created (instance methods)
        
    # A method to return a formatted string with showing date joined.
    def show_datejoined(self):
        return f"{self.fullname} joined on {self.date_joined:%m/%d/%y}"
# The class ends at the first un-indented line.
                
        
# Create an instance of the Member class named new_guy.
new_guy = Member('Rambo','Rocco Moe')
        
# Try out the date_joined method.
print(new_guy.show_datejoined())

Rocco Moe joined on 03/23/23


# Passing parameters to methods

In [43]:
# Method to activate (True) or deactivate (False) account.
def activate(self, yesno):
    """True for active, False to make inactive """
    self.is_active = yesno

#### It helps to understand that a method is really just a function. What makes a method different from a generic function is the fact that a method is always associated with some class. So it’s not as generic as a function.

# Adding and testing an .activate() method

In [47]:
import datetime as dt

# Define a new member class named Member.
class Member:
    def __init__(self, username, fullname):
        # Define attributes and give them values
        self.username = username
        self.fullname = fullname
        # Default date_joined to today's date
        self.date_joined = dt.date.today()
        # Set is active to True initially
        self.is_active = True
        
    # Methods for each instance created (instance methods)
    
    # A method to return a formatted string with showing date joined.
    def show_datejoined(self):
        return f"{self.fullname} joined on {self.date_joined:%m/%d/%y}"
    
    # Method to activate (True) or deactivate (False) account.
    def activate(self, yesno):
        """ True for active, False to make inactive """
        self.is_active = yesno
        
# The class ends at the first un-indented line.

# Create an instance of the Member class named new_guy.
new_guy = Member('Rambo','Rocco Moe')

# Is new_guy active?
print(new_guy.is_active)

# Try out the activate method.
new_guy.activate(False)

# Is new_guy still active?
print(new_guy.is_active)

# Print today's date
print(new_guy.date_joined)

True
False
2023-03-23


# Calling a class method by class name

# Using class variables

### Define the class variable

### This will give an error

# The variable free days is a class variable in this class

In [50]:
import datetime as dt
# Define a new class member name Member.
class Member:
    # Default number of free days.
    free_days = 365
    
    """ Create a new member """
    # This is an instance method for creating an object. Self immediately ties the method to one object
    def __init__(self, username, fullname): 
        self.date_joined = dt.date.today()
        # Set an expiration date
        self.free_expires = dt.date.today() + dt.timedelta(days = Member.free_days)
        
# The class ends at the first un-indented line.

# Create an instance of the Member class named new_guy.
wilbur = Member('wblomgren','Wilbur Blomegren')

print(wilbur.date_joined)
print(wilbur.free_expires)

2023-03-23
2024-03-22


In [51]:
# Set a default for free days
Member.free_days = 90

# Using class methods

#### The @ at the start of this defines it as a decorator — yep, yet another buzzword to add to your ever-growing list of nerd-o-rama buzzwords. A decorator is generally something that alters or extends the functionality of that to which it is applied.

In [62]:
import datetime as dt
# Define a new class name Member.
class Member:
    # Default number of free days.
    free_days = 365
    
    """ Create a new member. """
    def __init__(self, username, fullname):
        self.date_joined = dt.date.today()
        # Set an expiration date
        self.free_expires = dt.date.today() + dt.timedelta(days = Member.free_days)
        
    # Class methods follows @classmethod decorator and refer to cls rather than to self.
    @classmethod
    def setfreedays(cls,days):
        cls.free_days = days

In [63]:
Member.setfreedays(30)

In [64]:
wilbur = Member('wblomgren','Wilbur Blomgren')
print(wilbur.date_joined)
print(wilbur.free_expires)

2023-03-23
2023-04-22


# Using static methods

### What makes a static method different from instance and class methods is that a static method doesn’t relate specifically to an instance

# The Member class now has a static method named currenttime()

In [3]:
import datetime as dt

# Define a class named Member for making member objects.
class Member:
    # This is the class variable that's the same for all instances.
    free_days = 0
    
    """ Create a member object from username and fullname """
    def __init__(self, username, fullname):
        # Define properties and assign default values.
        self.datejoined = dt.date.today()
        self.free_expires = dt.date.today() + dt.timedelta(Member.free_days)
        
    
    # Class methods follow @classmethods and use cls rather than self.
    @classmethod
    def setfreedays(cls,days):
        cls.free_days = days
        
        
    @staticmethod
    def currenttime():
        now = dt.datetime.now()
        return f"{now:%I:%M %p}"
    
# Class definition ends at last indented line

# Try out the new static method (no object required)
print(Member.currenttime())

12:14 PM


# Understanding Class Inheritance

In [5]:
import datetime as dt
# Class is used for all kinds of people.
import datetime as dt

# Base class is used for all kinds of Members
class Member:
    """ The Member class attributes and methods are for everyone """
    # By default, a new account expires in one year (365 days)
    expiry_days = 365
    
    # Initialize a member object.
    def __init__(self, firstname, lastname):
        # attributes (instance variables) for everybody.
        self.firstname = firstname
        self.lastname = lastname
        # Calculate expiry date from today's date.
        self.expiry_date = dt.date.today() + dt.timedelta(days=self.expiry_days)

# Changing the value of an attribute

Syntax:

# Defining attributes with default values

Use the syntax:

# Changing the value of an object's attribute

In [12]:
# Define a new class name Member.
class Member:
    """ Create a new member. """
    def __init__(self, username, fullname):
        # Define attributes and give them values.
        self.username = username
        self.fullname = fullname
        
# The class ends at the instance, as well as its individual properties.

# Create an instance of the Member class named new_guy
new_guy = Member('Rambo','Rocco Moe')

# See what's in the instance, as well as its individual properties
print(new_guy.username)
print(new_guy.fullname)
print() # This just prints a blanks line.

# Change new_guy's user name then print both attributes again.
new_guy.username = "Princess"
print(new_guy.username)
print(new_guy.fullname)

Rambo
Rocco Moe

Princess
Rocco Moe


In [14]:
import datetime as dt

# Define a new class name Member.
class Member:
    """ Create a new member. """
    def __init__(self, username, fullname):
        # Define attributes and give them values.
        self.username = username
        self.fullname = fullname
        
        # Default date_joined to today's date.
        self.date_joined = dt.date.today()
        # Set is active to True initially
        self.is_active = True

# Giving a Class Methods

Use this syntax:

In [15]:
# Define methoda as functions, use self for "this" object.
def show_datejoined(self):
    return f"{self.fullname} joined on {self.date_joined:%m/%d/%y}"

# Changing the value of an object's attributes

In [17]:
import datetime as dt

# Define a new class name Member.
class Member:
    """ Create a new member """
    def __init__(self, username, fullname):
        # Define attributes and give them values
        self.username = username
        self.fullname = fullname
        # Define date_joined to today's date.
        self.date_joined = dt.date.today()
        # Set is active to True initially.
        self.isActive = True
    
    # Methods for each instance created (instance methods)
    
    # A method to return a formatted string with showing date joined.
    def show_datejoined(self):
        return f"{self.fullname} joined on {self.date_joined:%m/%d/%y}"
    
# The class ends at the first un-indented line.

# Create an instance of the Member class named new_guy
new_guy = Member('Rambo','Rocco Moe')

# Try out the date_joined method.
print(new_guy.show_datejoined())


Rocco Moe joined on 03/28/23


# Passing parameters to methods

In [18]:
# Method to activate (True) or deactivate (False) account.
def activate(self, yesno):
    """ True for active, False to make incative """
    self.is_active = yesno

# Adding and testing an .activate() method

In [22]:
import datetime as dt

# Define a new class name Member.
class Member:
    """ Create a new member """
    def __init__(self, username, fullname):
        # Define attributes and give them values.
        self.username = username
        self.fullname = fullname
        # Default date_joined to today's date.
        self.date_joined = dt.date.today()
        # Set is active to True initially.
        self.is_active = True
        
    # Methods for each instance created (instance methods)
    
    # A method to return a formatted str9ing with showing date joined.
    def show_datejoined(self):
        return f"{self.fullname} joined on {self.date_joined:%m/%d/%y}"
    
    def activate(self, yesno):
        """ True for active, False to make inactive """
        self.is_active = yesno
    
# The class ends at the first un-indented line.

# Create an instance of the Member class named new_guy.
new_guy = Member('Rambo','Rocco Moe')

# Is new-guy active?
print(new_guy.is_active)


# Try out the date_joined method.
new_guy.activate(False)

# Is new-guy still active?
print(new_guy.is_active)

True
False


# Calling a class method by class name

You can call a class method using the following syntax:

Alternative syntax is:

Create a new member using the code

# Using class variables

Using the standard syntax:

This wouldn’t work. If you tried to run the code that way, you’d get an error
saying Python doesn’t recognize the free_days variable name (even though it’s
defined right at the top of the class).

In [27]:
import datetime as dt
# Define a new class name Member.
class Member:
    # Default number of free days.
    free_days = 365
    
    """ Create a new member. """
    def __init__(self, username, fullname):
        self.date_joined = dt.date.today()
        # Set an expiration date
        self.free_expires = dt.date.today() + dt.timedelta(days = Member.free_days)
       
        
# The class ends at the first un-indented line.

# Create an instance of the Member class named new_guy.
wilbur = Member('wblomgren','Wilbur Blomegren')

print(wilbur.date_joined)
print(wilbur.free_expires)

2023-03-28
2024-03-27


So, what if later down the road you decide giving people 90 free days is plenty. You
could just change that in the class directly, but since it’s a variable, you can do it
on-the-fly, like this, outside the class:

# Using class methods

Syntax:

Define your class methods with this syntax:

In [28]:
# Class methods follow @classmethods and cls rather then self.
@classmethod
def setfreedays(cls,days):
    cls.free_days = days

# The set freedays() method is a class method in this class

In [29]:
import datetime as dt
# Define a new class name Member.
class Member:
    free_days = 365
    
    """ Create a new member """
    def __init__(self, username, fullname):
        self.date_joined = dt.date.today()
        # Set an expiration date
        self.free_expires = dt.date.today() + dt.timedelta(days = Member.free_days)
        
    # Class methods follow @classmethod decorator and refer to cls rather than to self.
    @classmethod
    def setfreedays(cls,days):
        cls.free_days = days

In [30]:
Member.setfreedays(30)

# Using static methods

Syntax:

In [32]:
@staticmethod
def currenttme():
    now = dt.datetime.now()
    return f"{now:%I:%M %p}"

# The Member class now has a static method named currenttime()

In [33]:
import datetime as dt

# Define a class named Member for making member objects.
class Member:
    # This is a class variable that's the same for all instances.
    free_days = 0
    
    """ Create a member object from username and fullname """
    def __init__(self, username, fullname):
        # Define properties and assign default values.
        self.datejoined = dt.date.today()
        self.free_expires = dt.date.today() + dt.timedelta(Member.free_days)
        
    # Class methods follow @classmethods and use cls rather than self.
    @classmethod
    def setfreedays(cls,days):
        cls.free_days = days
        
    @staticmethod
    def currenttime():
        now = dt.datetime.now()
        return f"{now:%I:%M %p}"
    
# Class definition ends at last indented line

# Try out the static method (no object required)
print(Member.currenttime())

04:23 PM


# Understanding Class Inheritance

### Creating the base (main) class

In [1]:
import datetime as dt
# Class is used for all kinds of people.
import datetime as dt

# Base class is used for all kinds of Members.
class Member:
    """ The Member class attributes and methods are everyone """
    # By default, a new account expires in one year (365 days)
    expiry_days = 365
    
    # Initialize a member object.
    def __init__(self, firstname, lastname):
        # Attributes (instance variables) for everybody.
        self.firstname = firstname
        self.lastname = lastname
        # Calculate expiry date from today'date.
        self.expiry_day = dt.date.today() + dt.timedelta(days=self.expiry_days)

# A simplified member class

In [2]:
import datetime as dt
# Class is used for all kinds of people.
import datetime as dt

# Base class is used for all kinds of Members.
class Member:
    """ The Member class attributes and methods are for everyone """
    # By default, a new account expires in one year (365 days)
    expiry_days = 365
    
    # Initialize a member object.
    def __init__(self, firstname, lastname):
        # Attributes (instance variables) for everybody.
        self.firstname = firstname
        self.lastname = lastname
        # Calculate expiry date from today's date.
        self.expiry_date = dt.date.today() + dt.timedelta(days=self.expiry_days)

# Outside the class now.
Joe = Member('Joe','Anybody')
print(Joe.firstname)
print(Joe.lastname)
print(Joe.expiry_date)

Joe
Anybody
2024-03-29


# Defining a subclass

Syntax:

In [ ]:
class Admin(Person):
    

In [4]:
# Subclass for Admins.
class Admin(Member):
    pass

In [5]:
# Subclass for Users.
class User(Member):
    pass

In [6]:
Ann = Admin('Annie','Angst')

In [7]:
Uli = User('Uli','ungula')

# Creating and Testing a Person subclass

In [8]:
# Subclass for Admin.
class Admin(Member):
    pass

# Subclass for Users.
class User(Member):
    pass

Ann = Admin('Annie', 'Angst')
print(Ann.firstname)
print(Ann.lastname)
print(Ann.expiry_date)
print()
Uli = User('Uli','Ungula')
print(Uli.firstname)
print(Uli.lastname)
print(Uli.expiry_date)

Annie
Angst
2024-03-29

Uli
Ungula
2024-03-29


# Overriding a default value from a subclass

In [9]:
# Subclass for admins.
class Admin(Member):
    # Admin accounts don't expire for 100 years.
    expiry_days = 365.2422 * 100

# Adding extra parameters from a subclass

# The Admin subclass has a new secret_code parameter

In [13]:
# Subclass for Admins.
class Admin(Member):
    # Admin accounts don't expire for 100 years.
    expiry_days = 365.2422 * 100
    # Subclass parameters
    def __init__(self, firstname, lastname, secret_code):
        # Pass Member parameters on up to Member class.
        super().__init__(firstname, lastname)
        # Assign the remaining parameter to this object.
        self.secret_code = secret_code
        
# Subclas for Users.
class User(Member):
    pass

Ann = Admin('Annie','Angst','PRESTO')
print(Ann.firstname, Ann.lastname, Ann.expiry_date, Ann.secret_code)

print()
Uli = User('Uli','Ungula')
print(Uli.firstname, Uli.lastname, Uli.expiry_date)

Annie Angst 2123-03-30 PRESTO

Uli Ungula 2024-03-29


 # The complete Admin and User subclass

In [16]:
import datetime as dt

# Base class is used all kinds of Members
class Member:
    """ The Member class properties and methods are for everyone """
    # By default, a new account expires in one year (365 days)
    expiry_days = 365
    
    # Initialize a member object.
    def __init__(self, firstname, lastname):
        # Properties (instance variables) for everybody.
        self.firstname = firstname
        self.lastname = lastname
        # Calculate expiry date from today's date
        self.expiry_date = dt.date.today() + dt.timedelta(days=self.expiry_days)
        # Default secret code is nothing
        self.secret_code = ''
        
    # Method in the base class 
    def showexpiry(self):
        return f"{self.firstname} {self.lastname} expires on {self.expiry_date}"
    
# Subclass for Admins.
class Admin(Member):
    # Admin accounts don't expire for 100 years.
    expiry_days = 365.2422 * 100

    # Subclass parameters
    def __init__(self, firstname, lastname, secret_code):
        # Pass Member parameters on up to Member class.
        super().__init__(firstname, lastname)
        # Assign the remaining parameter to this object.
        self.secret_code = secret_code
            
# Subclass for Users.
class User(Member):
    pass

Ann = Admin('Annie','Angst','PRESTO')
print(Ann.firstname, Ann.lastname, Ann.expiry_date, Ann.secret_code)
print() # Add a blank line to output.

Uli = User('Uli','Ungula')
print(Uli.firstname, Uli.lastname, Uli.expiry_date, Uli.secret_code)
        

Annie Angst 2123-03-30 PRESTO

Uli Ungula 2024-03-29 


# Calling a base class method

In [17]:
class Member:
    """ The Member class attributes and methods are for everyone """
    # By default, a new account expires in one year (365 days)
    expiry_days = 365
    
    # Initialize a member object.
    def __init__(self, firname, lastname):
        # Attributes (instance variables) for everybody.
        self.firname = firname
        self.lastname = lastname
        # Calculate expiry date from today's date.
        self.expiry_date = dt.date.today() + dt.timedelta(days=self.expiry_days)
        # Default secret_code = ''
        
    # Method in the base class
    def showexpiry(self):
        return f"{self.firname}{self.lastname} expires on {self.expiry_date}"

In [18]:
Ann = Admin('Annie','Angst','PRESTO')
print(Ann.showexpiry())

Uli = User('Uli','Ungula')
print(Uli.showexpiry())

Annie Angst expires on 2123-03-30
Uli Ungula expires on 2024-03-29


# Using the same name twice

In [19]:
class Member:
    """ The Member class attributes and methods """
    # Initialize a member object.
    def __init__(self, firstname, lastname):
        # Attributes (instance variables) for everybody.
        self.firstname = firstname
        self.lastname = lastname
        
    # Method in the base class
    def get_status(self):
        return f"{self.firstname} is a Member."
    
# Subclass for Adminstrators
class Admin(Member):
    def get_status(self):
        return f"{self.firstname} is an Admin."
    
# Subclass for regular Users
class User(Member):
    def get_status(self):
        return f"{self.firstname} is a regular User."

Python has a built-in help() method that you can use with any class to get more
information about that class. For example, at the bottom of the code in Figure 6-19,
add this line:

In [20]:
help(Admin)

Help on class Admin in module __main__:

class Admin(Member)
 |  Admin(firstname, lastname)
 |  
 |  Method resolution order:
 |      Admin
 |      Member
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  get_status(self)
 |  
 |  ----------------------------------------------------------------------
 |  Methods inherited from Member:
 |  
 |  __init__(self, firstname, lastname)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  ----------------------------------------------------------------------
 |  Data descriptors inherited from Member:
 |  
 |  __dict__
 |      dictionary for instance variables (if defined)
 |  
 |  __weakref__
 |      list of weak references to the object (if defined)



# Three methods with the same name, get_status()

In [24]:
class Member:
    """ The Member class attributes and methods are for everyone """
    # Initialize a member object.
    def __init__(self, firstname, lastname):
        # Attributes (instance variables) for everybody.
        self.firstname = firstname
        self.lastname = lastname
        
    # Method in the main class
    def get_status(self):
        return f"{self.firstname} is a Member."
    
# Subclass for Adminstrators
class Admin(Member):
    def get_status(self):
        return f"{self.firstname} is a Member."
    
# Subclass for regular Users
class Users(Member):
    def get_status(self):
        return f"{self.firstname} is a Member."
    
# Create an Admin
Ann = Admin('Annie','Angst')
print(Ann.get_status())

# Create a User
Uli = User('Uli','Ungula')
print(Uli.get_status)

# Create a member (neither Admin or User)
Manny = Member("Mindy","Membo")
print(Manny.get_status)

Annie is a Member.
<bound method User.get_status of <__main__.User object at 0x7f1323967dc0>>
<bound method Member.get_status of <__main__.Member object at 0x7f13235e54b0>>


In [25]:
help(Admin)

Help on class Admin in module __main__:

class Admin(Member)
 |  Admin(firstname, lastname)
 |  
 |  Method resolution order:
 |      Admin
 |      Member
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  get_status(self)
 |  
 |  ----------------------------------------------------------------------
 |  Methods inherited from Member:
 |  
 |  __init__(self, firstname, lastname)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  ----------------------------------------------------------------------
 |  Data descriptors inherited from Member:
 |  
 |  __dict__
 |      dictionary for instance variables (if defined)
 |  
 |  __weakref__
 |      list of weak references to the object (if defined)



In [26]:
print(Admin.__dict__)

{'__module__': '__main__', 'get_status': <function Admin.get_status at 0x7f132377a950>, '__doc__': None}


# Chapter 7 
# Sidestepping Errors

## Understanding Exceptions

The Syntax is:

## Handling Errors Gracefully

Basic Syntax:

Here’s how we can rewrite the showfilecontents.py code to handle the missing
(or misspelled) file error:

In [27]:
try:
    # Open file and shows its name.
    thefile = open('people.csv')
    print(thefile.name)
except Exception:
    print("Sorry i don't see a file named people.csv here")

Sorry i don't see a file named people.csv here


# Basic Specific about Exceptions

In [28]:
try:
    # Open file and show its name.
    thefile = open('people.csv')
    print(thefile.name)
    print(thefile.wookens)
except Exception:
    print("Sorry, I don't see a file named people.csv here")

Sorry, I don't see a file named people.csv here


### The correct way to show or display the error

In [30]:
try:
    # Open file and shows its name.
    thefile = open('people.csv')
    print(thefile.name)
    print(thefile.wookems())
except FileNotFoundError:
    print("Sorry, I don't see a file named people.csv here")

Sorry, I don't see a file named people.csv here


# Keeping your app from crashing 

In [31]:
try:
    # Open file and its name.
    thefile = open('people.csv')
    # Print a couple blank lines then the first line from the file.
    print('\n\n', thefile.readline())
    # Close the file.
    thefile.closed()
    
except FileNotFoundError:
    print("Sorry, I don't see a file named people.csv here")
except Exception:
    print("Sorry, something else went wrong")

Sorry, I don't see a file named people.csv here


In [32]:
try:
    # Open file and show its name.
    thefile = open('people.csv')
    # Print a couple blank lines then the first line from the file.
    print('\n\n', thefile.readline())
    # Close the file.
    thefile.closed()
    
except FileNotFoundError:
    print("Sorry, I don't see a file named people.csv here")
except Exception:
    print("Sorry, something else went wrong")

Sorry, I don't see a file named people.csv here


For example, consider the following rewrite of the preceding code. The generic
handler, except Exception now has an as e at the end, which means “whatever
exception gets caught here, put the error message in a variable named e.” Then
the next line uses print(e) to display whatever is in that e variable:

In [33]:
try:
    # Open file and shows its name.
    thefile = open('people.csv')
    # Print a couple of blank lines then the first line from the file.
    print('\n\n', thefile.readline())
    thefile.wigwam()
    
except FileNotFoundError:
    print("Sorry, I don't see a file named people.csv here")
except Exception as e:
    print(e)

Sorry, I don't see a file named people.csv here


# Adding an else to the Mix

A more elegant way to deal with the problem uses this syntax:

Here is all the code with com-
ments explaining what’s going on:

# Code with try, exception handlers, and an else for when there are no exceptions

In [34]:
try:
    # Open the file name people.csv
    thefile = open('people.csv')
# Watch for common error and stop program if it happens.
except FileNotFoundError:
    print("Sorry, I don't see a file named people.csv here")
# Catch any unexpected error and stop program if one happens.
except Exception as err:
    print(err)
# Otherwise, if nothing bad has happened by now, just keep going.
else:
    # File must be open by now if we got here.
    print('\n') # Print a blank line.
    # Print each line from the file.
    for one_line in thefile:
        print(one_line)
    thefile.close()
    print("Success!")

Sorry, I don't see a file named people.csv here


If you look at the complete syntax for Python exception handling, you’ll see there
is one more option at the end, like this:

# To illustrate, here is some code that expects to find an external resource named people.csv to be available to the code.

In [35]:
print('Do this first')
try:
    open('people.csv')
except FileNotFoundError:
    print('Cannot find file named people.csv')
except Exception as e:
    print(e)
else:
    print('Show this if no exception.')
finally:
    print('This is in the finally block')
print("This is outside the try...except...else...finally")


Do this first
Cannot find file named people.csv
This is in the finally block
This is outside the try...except...else...finally


# Raising Your Own Errors

The general syntax for raising your own error is:

In [1]:
try:
    # Open the file (no error check for this example)
    thefile = open('people.csv')
    # Count the number of lines in the file.
    line_count = len(thefile.readlines())
    # If there few than two lines, raise exception.
    if line_count < 2:
        raise Exception
    # Handles missing file error.
except FileNotFoundError:
    print('\nThere is no people.csv file here')


There is no people.csv file here


# Custom EmptyFileError added for exception handling

In [3]:
# Base class for defining your own user-defined exceptions.
class Error(Exception):
    """ Base class for other Exception """
    pass

# Now defined your exception as a subclass of error.
class EmptyFileError(Error):
    pass

try:
    # Open the file (no error check for this example).
    thefile = open('people.csv')
    # If there is fewer than 2 lines, raise exception.
    if line_count < 2:
        raise EmptyFileError
        
# Handles missing file error.
except FileNotFoundError:
    print("\nYour people.csv file here")
    
# Handles my custom error for too few rows.
except EmptyFileError:
    print("\nYour people.csv file doesnt have enough stuff.")
    
# Handles all other exceptions
except Exception as e:
    # Show the error.
    print('\n\nFailed: The error was ' + str(e))
    # close the file.
    thefile.close()
else:
    # This code runs only if no exceptions above.
    print() # Print a blank line.
    
    # File must be open by now if we got here, show content.
    for one_line in thefile:
        print(one_line)
    thefile.close()
    print("Success!")


Your people.csv file here


# Book 3 
# Working with Python Libraries

# In this chapter we will use VS Code.................................

# Openning and Closing Files

To open a file from within a Python app, use the syntax:

There are basically two ways to use the open method. With one syntax you assign
a variable name to the file, and use this variable name in later code to refer to the
file. That syntax is:

For now, we simply copy everything that’s in the file to
a variable named filecontents in Python, and then we display this content using
a simple print() function. So to open quotes.txt, read in all its content, and
display that content on the screen, use this code:

With this method, the file remains open until you specifically close it using the file
variable name and the .close() method, like this:

The very last thing on the line is a colon which marks the beginning of the
with block. All indented code below that is assumed to be relevant to the context
of the open file (like code indented inside a loop). At the end of this you don’t need
to specifically close the file; Python does it automatically:

The following code shows a single app that opens quotes.txt, reads and displays
its content, and then closes the file. With the first method you have to specifically
use .close() to close the file. With the second, the file closes automatically, so no
.close() is required:

The previous example works fine because quotes.txt is a really simple text file
that contains only ASCII characters — the kinds of letters, numbers, and punctua-
tion marks that you can type from a standard keyboard for the English language.

Attempting to run this code results in the following error:

This isn’t the most helpful message in the world. Suppose you try to open names.
txt, which (one would assume) is a text file like quotes.txt, using this code:

You run this code, and again you get a strange error message, like this:

So what the heck is going on here?
The first problem is caused by the fact that the file is a .jpg, a graphic image,
which means it’s a binary file, not a text file. So to open this one, you need a b in
the mode. Or use rb, which means read binary, like this:

Running this code doesn’t generate an error. But what it does show doesn’t look
anything like the actual picture. The output from this code is a lot of stuff that
looks something like this: